src/ring_theory/ideal_over.lean

```diff
@@ -47,28 +48,93 @@ begin
+  ∃ i, p.coeff i ∈ (J.comap f : set R) \ (I.comap f) :=
+begin
+  obtain ⟨hrJ, hrI⟩ := hr,
+  have rbar_ne_zero : quotient.mk I r ≠ 0 := mt (quotient.mk_eq_zero I).mp hrI,
+  have rbar_mem_J : quotient.mk I r ∈ J.map (quotient.mk I) := mem_map_of_mem hrJ,
+  have quotient_f : ∀ x ∈ I.comap f, (quotient.mk I).comp f x = 0,
+  { simp [quotient.eq_zero_iff_mem] },
+  have rbar_root : (p.map (quotient.mk (I.comap f))).eval₂
+    (quotient.lift (I.comap f) _ quotient_f)
+    (quotient.mk I r) = 0,
+  { convert quotient.eq_zero_iff_mem.mpr hpI,
+    exact trans (eval₂_map _ _ _) (hom_eval₂ p f (quotient.mk I) r).symm },
+  obtain ⟨i, ne_zero, mem⟩ :=
+    exists_coeff_ne_zero_mem_comap_of_root_mem rbar_ne_zero rbar_mem_J p_ne_zero rbar_root,
+  rw coeff_map at ne_zero mem,
+  refine ⟨i, mem_of_mem_quotient hIJ _, mt _ ne_zero⟩,
+  { simpa using mem },
+  simp [quotient.eq_zero_iff_mem],
+end
+
 lemma comap_ne_bot_of_root_mem {r : S} (r_ne_zero : r ≠ 0) (hr : r ∈ I)
   {p : polynomial R} (p_ne_zero : p ≠ 0) (hp : p.eval₂ f r = 0) :
   I.comap f ≠ ⊥ :=
 λ h, let ⟨i, hi, mem⟩ := exists_coeff_ne_zero_mem_comap_of_root_mem r_ne_zero hr p_ne_zero hp in
 absurd ((mem_bot _).mp (eq_bot_iff.mp h mem)) hi
 
+lemma comap_lt_comap_of_root_mem_sdiff [I.is_prime] (hIJ : I ≤ J)
+  {r : S} (hr : r ∈ (J : set S) \ I)
+  {p : polynomial R} (p_ne_zero : p.map (quotient.mk (I.comap f)) ≠ 0) (hp : p.eval₂ f r ∈ I) :
+  I.comap f < J.comap f :=
+let ⟨i, hJ, hI⟩ := exists_coeff_mem_comap_sdiff_comap_of_root_mem_sdiff hIJ hr p_ne_zero hp
+in lt_iff_le_and_exists.mpr ⟨comap_mono hIJ, p.coeff i, hJ, hI⟩
+
 variables [algebra R S]
 
-lemma comap_ne_bot_of_algebraic_mem {I : ideal S} {x : S}
+lemma comap_ne_bot_of_algebraic_mem {x : S}
   (x_ne_zero : x ≠ 0) (x_mem : x ∈ I) (hx : is_algebraic R x) : I.comap (algebra_map R S) ≠ ⊥ :=
 let ⟨p, p_ne_zero, hp⟩ := hx
 in comap_ne_bot_of_root_mem x_ne_zero x_mem p_ne_zero hp
 
-lemma comap_ne_bot_of_integral_mem [nontrivial R] {I : ideal S} {x : S}
+lemma comap_ne_bot_of_integral_mem [nontrivial R] {x : S}
   (x_ne_zero : x ≠ 0) (x_mem : x ∈ I) (hx : is_integral R x) : I.comap (algebra_map R S) ≠ ⊥ :=
 comap_ne_bot_of_algebraic_mem x_ne_zero x_mem (hx.is_algebraic R)
 
+lemma mem_of_one_mem (h : (1 : S) ∈ I) (x) : x ∈ I :=
+(I.eq_top_iff_one.mpr h).symm ▸ mem_top
+
+lemma comap_lt_comap_of_integral_mem_sdiff [I.is_prime] (hIJ : I ≤ J)
+  {x : S} (mem : x ∈ (J : set S) \ I) (integral : is_integral R x) :
+  I.comap (algebra_map R S) < J.comap (algebra_map _ _) :=
+begin
+  obtain ⟨p, p_monic, hpx⟩ := integral,
+  refine comap_lt_comap_of_root_mem_sdiff hIJ mem _ _,
+  swap,
+  { intro h,
+    refine mem.2 (mem_of_one_mem _ _),
+    rw [←(algebra_map R S).map_one, ←mem_comap, ←quotient.eq_zero_iff_mem],
+    apply (map_monic_eq_zero_iff p_monic).mp h 1 },
+  convert I.zero_mem
+end
+
+lemma is_maximal_of_is_integral_of_is_maximal_comap
+  (hRS : ∀ (x : S), is_integral R x) (I : ideal S) [I.is_prime]
```

Comment:
Don't we have an abbreviation for that? I thought there was `algebra.is_integral R S`. But maybe only `algebra.is_algebraic` ...

Reply:
I cannot find any abbreviation, only the fully written out version [for example here](https://github.com/leanprover-community/mathlib/blob/master/src/ring_theory/integral_closure.lean#L312).

I am in favour of adding a predicate `ring_hom.is_integral` and refactoring existing files to use this instead of inferring it from the `algebra` instance, but that seems like it would be its own PR.